File: r1-spike/server/lib/loka_r1_server/scale.ex
defmodule LokaR1Server.Scale do
  @moduledoc """
  Quick R1-A3 timing on the server (owner decision 2026-09-24): the unchanged
  Elixir kernel behind the A2 store (`LokaR1Server.Store`: WAL, `synchronous=FULL`,
  raw `BEGIN IMMEDIATE`/`COMMIT`), at Tiny/Medium/Stress state size. The step is
  the same as `mobile/scale.ts`, so the two CSVs share one summarizer:

  - admission: the receipt row for this request id, and the kernel HOST value
    (memory, that one receipt if any, no pending, empty published);
  - decision: `LokaR1.World.step/3`, in process (no boundary encode/decode);
  - encode: `Codec.encode` of the durable state (only when the revision changed),
    the new receipt and the new events;
  - commit: `BEGIN IMMEDIATE`, UPDATE durable, INSERT receipt, `COMMIT`;
  - projection: adopt the new memory and encode `{result, events, delta}`.

  It runs in the calling process, not behind the A2 host's GenServer call, and
  the outbox isn't durable on the server (as in A2), so events are encoded but
  not written.
  """
  alias LokaR1.{Codec, World}
  alias LokaR1Server.Store

  @header "host,model,i,action,outcome,code,state_write,admission_ms,decision_ms,encode_ms,commit_ms,projection_ms,e2e_ms"
  @checkpoints 20

  @spec header() :: String.t()
  def header, do: @header

  @doc "Runs one model in a fresh database at `db`; returns `{csv_rows, facts}`."
  @spec run_model(Path.t(), map(), String.t()) :: {[String.t()], map()}
  def run_model(db, input, host_name) do
    %{"model" => model, "world" => world_name, "initial" => initial} = input
    {decide, _} = World.world(world_name)
    conn = Store.open(db)
    id = "a3-" <> model
    :ok = Store.create(conn, id, world_name, initial)

    Store.exec!(
      conn,
      "CREATE TABLE IF NOT EXISTS checkpoint (k TEXT PRIMARY KEY, v TEXT NOT NULL)"
    )

    pragmas = %{
      "journal_mode" => hd(hd(Store.exec!(conn, "PRAGMA journal_mode"))),
      "synchronous" => hd(hd(Store.exec!(conn, "PRAGMA synchronous")))
    }

    {_, memory, nil} = Store.instance(conn, id)
    warmup = input["warmup"]

    {rows, memory} =
      input["commands"]
      |> Enum.with_index()
      |> Enum.flat_map_reduce(memory, fn {command, i}, memory ->
        {row, memory} = step(conn, id, decide, memory, command)

        if rem(i + 1, 250) == 0,
          do: IO.puts(:stderr, "LOKA_A3_PROGRESS #{model} #{i + 1}/#{length(input["commands"])}")

        row = [host_name, model, i | row] |> Enum.join(",")
        {if(i >= warmup, do: [row], else: []), memory}
      end)

    pre = Codec.encode(memory)

    checkpoints =
      for _ <- 1..@checkpoints do
        t0 = now()
        Store.exec!(conn, "BEGIN IMMEDIATE")
        Store.exec!(conn, "INSERT OR REPLACE INTO checkpoint VALUES ('state', ?1)", [pre])
        Store.exec!(conn, "COMMIT")
        [[text]] = Store.exec!(conn, "SELECT v FROM checkpoint WHERE k = 'state'")
        {:ok, value} = Codec.decode(text)
        back = Codec.encode(value)
        t1 = now()
        back == pre || raise "#{model}: checkpoint round trip changed the canonical state"
        round((t1 - t0) / 1000)
      end

    [[durable]] = Store.exec!(conn, "SELECT durable FROM instances WHERE id = ?1", [id])
    durable == pre || raise "#{model}: durable row differs from memory at the end"
    [[receipts]] = Store.exec!(conn, "SELECT count(*) FROM receipts WHERE instance = ?1", [id])
    Store.close(conn)

    facts = %{
      "model" => model,
      "version" => input["version"],
      "seed" => input["seed"],
      "steps" => length(input["commands"]),
      "warmup" => warmup,
      "initial_state_bytes" => byte_size(Codec.encode(initial)),
      "final_state_bytes" => byte_size(pre),
      "final_state_sha256" => Base.encode16(:crypto.hash(:sha256, pre), case: :lower),
      "final_narration_entries" => length(memory["narration"]),
      "receipts" => receipts,
      "sqlite" => pragmas,
      "checkpoint_round_trip_us" => checkpoints
    }

    {rows, facts}
  end

  defp step(conn, id, decide, memory, %{"request" => %{"id" => rid} = request} = command) do
    t0 = now()
    prior = Store.receipt(conn, id, rid)
    receipts = if prior, do: %{rid => prior}, else: %{}

    host = %{
      "memory" => memory,
      "durable" => memory,
      "receipts" => receipts,
      "pending" => nil,
      "in_doubt" => false,
      "published" => []
    }

    t1 = now()
    {record, after_host} = World.step(decide, host, command)
    t2 = now()
    fresh = prior == nil and Map.has_key?(after_host["receipts"], rid)
    durable = after_host["durable"]
    state_write = fresh and durable["revision"] != memory["revision"]
    durable_text = if state_write, do: Codec.encode(durable)
    receipt = if fresh, do: after_host["receipts"][rid]
    receipt_result = if fresh, do: Codec.encode(receipt["result"])
    _events = Enum.map(after_host["published"], &Codec.encode/1)
    t3 = now()

    if fresh do
      Store.exec!(conn, "BEGIN IMMEDIATE")

      if durable_text,
        do:
          Store.exec!(conn, "UPDATE instances SET durable = ?2 WHERE id = ?1", [id, durable_text])

      Store.exec!(
        conn,
        "INSERT INTO receipts (instance, id, intent, result) VALUES (?1, ?2, ?3, ?4)",
        [id, rid, receipt["intent"], receipt_result]
      )

      Store.exec!(conn, "COMMIT")
    end

    t4 = now()
    memory = after_host["memory"]
    _response = Codec.encode(Map.take(record, ~w(result events delta)))
    t5 = now()
    result = record["result"]
    outcome = if result["delivery"] == "replay", do: "replayed", else: result["kind"]

    row = [
      request["action"],
      outcome,
      result["code"],
      if(state_write, do: 1, else: 0)
      | Enum.map([t1 - t0, t2 - t1, t3 - t2, t4 - t3, t5 - t4, t5 - t0], &ms/1)
    ]

    {row, memory}
  end

  defp now, do: System.monotonic_time(:nanosecond)
  defp ms(ns), do: :erlang.float_to_binary(ns / 1_000_000, decimals: 4)
end

File: r1-spike/mobile/scale.ts
// Quick R1-A3 timing (owner decision 2026-09-24): the unchanged TypeScript kernel
// behind a durable SQLite host, at Tiny/Medium/Stress state size. Platform-free:
// App.tsx supplies expo-sqlite, performance.now() and file output on the phone;
// scripts/scale-local.mjs runs the same code on node:sqlite (a pre-phone check, not evidence).
//
// Per step (one fresh `invoke` command, no faults):
//   admission   the receipt row for this request id (SELECT by primary key), and the kernel HOST
//               value built from process memory: memory, that one receipt (if any), no pending,
//               empty published. Receipts are host storage, not world state, so only the
//               admission-relevant one crosses into the kernel.
//   decision    kernel `step` (world.ts): whole-HOST clone, envelope checks, decide, delta, record.
//               In-process on Hermes: there is no further boundary encode/decode.
//   encode      canonical encoding of the durable state (only when the revision changed), the new
//               receipt and the new events.
//   commit      BEGIN IMMEDIATE; UPDATE durable; INSERT receipt; INSERT outbox events; COMMIT.
//   projection  adopt the new memory and encode the player response {result, events, delta}.
//   e2e         admission + decision + encode + commit + projection.
import { canonical, has, obj, parse } from '../ts/src/kernel/codec.ts';
import type { Json, JsonObject } from '../ts/src/kernel/codec.ts';
import { sha256Hex } from '../ts/src/kernel/sha256.ts';
import { utf8 } from '../ts/src/kernel/codec.ts';
import { WORLDS, step, validMemory } from '../ts/src/kernel/world.ts';
import type { Host } from '../ts/src/kernel/world.ts';
import type { Db } from './host.ts';

export type ScaleEnv = {
  host: string;
  open(name: string): Db;
  now(): number; // monotonic milliseconds
  write(name: string, text: string): void;
  log(line: string): void;
};

export const CSV_HEADER = 'host,model,i,action,outcome,code,state_write,admission_ms,decision_ms,encode_ms,commit_ms,projection_ms,e2e_ms';
const CHECKPOINTS = 20;

const SCHEMA = `
CREATE TABLE IF NOT EXISTS kv(k TEXT PRIMARY KEY, v TEXT NOT NULL);
CREATE TABLE IF NOT EXISTS receipts(id TEXT PRIMARY KEY, receipt TEXT NOT NULL);
CREATE TABLE IF NOT EXISTS outbox(seq INTEGER PRIMARY KEY, event TEXT NOT NULL);
CREATE TABLE IF NOT EXISTS checkpoint(k TEXT PRIMARY KEY, v TEXT NOT NULL);`;

const f = (ms: number): string => ms.toFixed(4);

/** Runs one model; returns its CSV rows (measured steps only) and summary facts. */
export function runModel(env: ScaleEnv, input: JsonObject): { rows: string[]; facts: JsonObject } {
  const model = input.model as string;
  const world = WORLDS[input.world as string]();
  const initial = input.initial as JsonObject;
  if (!validMemory(world, initial)) throw new Error(model + ': the kernel rejects the padded initial memory');
  const db = env.open('a3-' + model + '.db');
  db.exec(SCHEMA);
  db.exec('BEGIN IMMEDIATE');
  db.exec('DELETE FROM kv; DELETE FROM receipts; DELETE FROM outbox; DELETE FROM checkpoint');
  db.all('INSERT INTO kv VALUES (?, ?)', 'durable', canonical(initial));
  db.exec('COMMIT');
  const pragmas = { ...db.all('PRAGMA journal_mode')[0], ...db.all('PRAGMA synchronous')[0] };

  let memory = parse(db.all("SELECT v FROM kv WHERE k = 'durable'")[0].v as string) as JsonObject;
  const commands = input.commands as JsonObject[];
  const warmup = input.warmup as number;
  const rows: string[] = [];
  let outboxSeq = 0;
  for (let i = 0; i < commands.length; i++) {
    const command = commands[i];
    const request = command.request as JsonObject;
    const id = request.id as string;
    const t0 = env.now();
    const found = db.all('SELECT receipt FROM receipts WHERE id = ?', id);
    const receipts = obj();
    if (found.length) receipts[id] = parse(found[0].receipt as string);
    const host: Host = { memory, durable: memory, receipts, pending: null, in_doubt: false, published: [] };
    const t1 = env.now();
    const { record, host: after } = step(world, host, command);
    const t2 = env.now();
    const fresh = !found.length && has(after.receipts, id);
    const stateWrite = fresh && after.durable.revision !== memory.revision;
    const durableText = stateWrite ? canonical(after.durable) : null;
    const receiptText = fresh ? canonical(after.receipts[id]) : null;
    const eventTexts = after.published.map((e) => canonical(e));
    const t3 = env.now();
    if (fresh) {
      db.exec('BEGIN IMMEDIATE');
      if (durableText !== null) db.all("UPDATE kv SET v = ? WHERE k = 'durable'", durableText);
      db.all('INSERT INTO receipts VALUES (?, ?)', id, receiptText as string);
      for (const e of eventTexts) db.all('INSERT INTO outbox VALUES (?, ?)', ++outboxSeq, e);
      db.exec('COMMIT');
    }
    const t4 = env.now();
    memory = after.memory;
    const response = canonical(Object.assign(obj(), { result: record.result, events: record.events, delta: record.delta }));
    const t5 = env.now();
    const result = record.result as JsonObject;
    if (i >= warmup) {
      const outcome = result.delivery === 'replay' ? 'replayed' : (result.kind as string);
      rows.push([env.host, model, i, request.action, outcome, result.code, stateWrite ? 1 : 0,
        f(t1 - t0), f(t2 - t1), f(t3 - t2), f(t4 - t3), f(t5 - t4), f(t5 - t0)].join(','));
    }
    if (response.length === 0) throw new Error('empty response');
    if ((i + 1) % 250 === 0) env.log(`LOKA_A3_PROGRESS ${model} ${i + 1}/${commands.length}`);
  }

  // Checkpoint round trip: write the canonical state, read it back, parse, re-encode, compare.
  const pre = canonical(memory);
  const checkpoints: number[] = [];
  for (let k = 0; k < CHECKPOINTS; k++) {
    const t0 = env.now();
    db.exec('BEGIN IMMEDIATE');
    db.all('INSERT OR REPLACE INTO checkpoint VALUES (?, ?)', 'state', pre);
    db.exec('COMMIT');
    const back = canonical(parse(db.all("SELECT v FROM checkpoint WHERE k = 'state'")[0].v as string));
    const t1 = env.now();
    if (back !== pre) throw new Error(model + ': checkpoint round trip changed the canonical state');
    checkpoints.push(Math.round((t1 - t0) * 1000));
  }
  const durable = db.all("SELECT v FROM kv WHERE k = 'durable'")[0].v as string;
  if (durable !== pre) throw new Error(model + ': durable row differs from memory at the end');
  const facts = Object.assign(obj(), {
    model,
    version: input.version as Json,
    seed: input.seed as Json,
    steps: commands.length,
    warmup,
    initial_state_bytes: utf8(canonical(initial)).length,
    final_state_bytes: utf8(pre).length,
    final_state_sha256: sha256Hex(utf8(pre)),
    final_narration_entries: (memory.narration as Json[]).length,
    receipts: db.all('SELECT count(*) AS n FROM receipts')[0].n as number,
    sqlite: pragmas as JsonObject,
    checkpoint_round_trip_us: checkpoints,
  });
  return { rows, facts };
}

/** All models in order; writes a3-samples.csv then a3-run.json (its presence means done). */
export function runScale(env: ScaleEnv, inputLines: string[], extra: JsonObject): JsonObject {
  const rows: string[] = [CSV_HEADER];
  const models: Json[] = [];
  const start = env.now();
  for (const line of inputLines) {
    const r = runModel(env, parse(line) as JsonObject);
    rows.push(...r.rows);
    models.push(r.facts);
    env.log('LOKA_A3_MODEL ' + canonical(r.facts));
  }
  env.write('a3-samples.csv', rows.map((r) => r + '\n').join(''));
  const run = Object.assign(obj(), { host: env.host, wall_ms: Math.round(env.now() - start), models, ...extra });
  env.write('a3-run.json', canonical(run) + '\n');
  return run;
}
